keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -217,48 +219,52 @@ is not sensitive.
 The definition is as follows:
 ```
 type Credentials struct {
-  credentialProviders map[string]CredentialsConfig // mapping of credentials types to their config. In some cases the cluster may recognize different identity types and they may have different endpoints or TLS config.
+  // +listType=map
+  // +listMapKey=name
+  credentialProviders []CredentialsConfig // mapping of credentials types to their config. In some cases the cluster may recognize different identity types and they may have different endpoints or TLS config.
 }
 
 // CredentialsTypes defines the type of credentials that are accepted by the cluster. For example, GCP credentials (tokens that are understood by GCP's IAM) are designated by the string `google`.
 type CredentialsType string
 
 // CredentialsConfig gives more details on data that is necessary to reach out the cluster for this kind of Credentials
 type CredentialsConfig struct {
+  Name string // name of the provider type
   Cluster *Cluster // Configuration to reach the cluster (endpoints, proxy, etc) // See following section for details.
 }
 ```
 
 #### Cluster Data
 
 The Cluster structure for the exec defined in KEP 541, [implemented in k/k](https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/types.go#L73-L80) assumes the following:
+
 ```
 type Cluster struct {
 	// Server is the address of the kubernetes cluster (https://hostname:port).
-	Server string
+	Server string `json:"server"`
 	// TLSServerName is passed to the server for SNI and is used in the client to
 	// check server certificates against. If ServerName is empty, the hostname
 	// used to contact the server is used.
 	// +optional
-	TLSServerName string
+	TLSServerName string `json:"tls-server-name,omitempty"`
 	// InsecureSkipTLSVerify skips the validity check for the server's certificate.
 	// This will make your HTTPS connections insecure.
 	// +optional
-	InsecureSkipTLSVerify bool
+	InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
 	// CAData contains PEM-encoded certificate authority certificates.
 	// If empty, system roots should be used.
 	// +listType=atomic
```

Comment:
We cannot put this into CRD because the listType setting here.

Reply:
why?
```
// Example:
//
//	+listType=atomic
//
// Atomic lists will be entirely replaced when updated. This tag may be
// used on any type of list (struct, scalar, ...).
//
// Using this tag will generate the following OpenAPI extension:
//
//	"x-kubernetes-list-type": "atomic"
```

whats wrong with this? I think it applies only to `CertificateAuthorityData` field.

Reply:
ah bummer. maybe we could check with the original library, this may be a miss on their end?

Reply:
See the comment in the original file:

> To ensure that this struct contains everything someone would need to communicate
> with a kubernetes cluster (just like they would via a kubeconfig), the fields
> should shadow "k8s.io/client-go/tools/clientcmd/api/v1".Cluster, with the exception
> of CertificateAuthority, since CA data will always be passed to the plugin as bytes.

The data type exported to clients is https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1/types.go#L78 which doesn’t have `CertificateAuthorityData`.